pkg/querier/querier.go

```diff
@@ -321,34 +321,43 @@ func (q *Querier) Tail(ctx context.Context, req *logproto.TailRequest) (*Tailer,
 
 // passed to tailer for (re)connecting to new or disconnected ingesters
 func (q *Querier) tailDisconnectedIngesters(ctx context.Context, req *logproto.TailRequest, connectedIngestersAddr []string) (map[string]logproto.Querier_TailClient, error) {
-	tailClients := make(map[string]logproto.Querier_TailClient)
-	for i := range connectedIngestersAddr {
-		tailClients[connectedIngestersAddr[i]] = nil
+	// Build a map to easily check if an ingester address is already connected
+	connectedIngestersAddrs := make(map[string]bool)
+	for _, addr := range connectedIngestersAddr {
+		connectedIngestersAddrs[addr] = true
 	}
 
-	disconnectedIngesters := []ring.IngesterDesc{}
+	// Get the current replication set from the ring
 	replicationSet, err := q.ring.GetAll()
```

Comment:
`GetAll` returns all ingesters inside the ring that are either `ACTIVE` or `LEAVING`, but in this case we only want the `ACTIVE` ones; could you filter the list?

example:
 ```
// filterLiveIngesters returns a copy of a replication set where the set of ingesters
// only containers ingesters that are ACTIVE and not LEAVING
func filterLiveIngesters(rs ring.ReplicationSet) ring.ReplicationSet {
	ingesters := make([]ring.IngesterDesc, 0, len(rs.Ingesters))
	for _, ingester := range rs.Ingesters {
		if ingester.State != ring.ACTIVE {
			continue
		}
		ingesters = append(ingesters, ingester)
	}
	return ring.ReplicationSet{Ingesters: ingesters, MaxErrors: rs.MaxErrors}
}
```

Not filtering the list doesn't cause any crashes, but it will continuously get removed from the `connectedIngestersAddr` list until it finally leaves and clutters up the logs with `Ingester is stopping` messages..

Reply:
Very good point. I agree. What's about `PENDING` and `JOINING` states? Should they be filtered out like `LEAVING`?

Reply:
I think `JOINING` should be filtered out too since they're not ready to accept traffic. `PENDING` is already filtered out. Here's the code that does the filtering from Cortex (in this instance, Operation will be Read):

https://github.com/grafana/loki/blob/79333add88f28be425acee607cdb6fa612e8b5a4/vendor/github.com/cortexproject/cortex/pkg/ring/replication_strategy.go#L49-L57

I think in this case the only state we want is ACTIVE.

Reply:
_In the meanwhile, I've filtered out ingesters in the `LEAVING` state ([70b419a](https://github.com/grafana/loki/pull/897/commits/70b419ae9b264184a74f95967a1b2e5a61d5a6c7))._